chat.js

```diff
@@ -920,25 +926,55 @@ Chat.uncacheTree = function (root) {
 	} while (uncache.length > 0);
 };
 
-Chat.loadCommands = function () {
-	if (Chat.commands) return;
+Chat.loadPlugins = function () {
+	if (Chat.commands || Chat.filter) return;
 
 	FS('package.json').readTextIfExists().then(data => {
 		if (data) Chat.package = JSON.parse(data);
 	});
 
 	let baseCommands = Chat.baseCommands = require('./chat-commands').commands;
 	let commands = Chat.commands = Object.assign({}, baseCommands);
+	let chatfilters = Chat.filters;
 
-	// Install plug-in commands
+	// Install plug-in commands and chat filter
 
 	// info always goes first so other plugins can shadow it
 	Object.assign(commands, require('./chat-plugins/info').commands);
 
 	for (let file of FS('chat-plugins/').readdirSync()) {
 		if (file.substr(-3) !== '.js' || file === 'info.js') continue;
-		Object.assign(commands, require('./chat-plugins/' + file).commands);
+		const plugin = require(`./chat-plugins/${file}`);
+
+		Object.assign(commands, plugin.commands);
+
+		const filter = plugin.chatfilter;
+		if (filter) {
+			if (typeof filter !== 'function') {
+				require('./crashlogger')(new TypeError(`This chatfilter is not a function`), `Loading a chatfilter`, {
+					file: `File location: ../chat-plugins/${file}`,
+					filter: filter,
+					type: typeof filter,
+				});
+			} else {
+				chatfilters.push(filter);
+			}
+		}
 	}
+
+	Chat.filter = function (message, user, room, connection, targetUser) {
```

Comment:
Honestly, this should also be initialized at the top-level scope.

Reply:
I mean, I set it to ``undefined`` on the top-level scope.  I do that so that I can override it here, once other filters are loaded.

Reply:
There's no need. `Chat.filters` is globally accessible. Hotpatching chat replaces the entire `Chat` object, but even if it didn't, `Chat.filter` will always access the latest version of the `chatfilters` array.

Reply:
Much worse is if `Chat.filter` is called and it's undefined, because that will cause a crash.

Reply:
As a rule of thumb, functions pretty much _always_ need to be initialized immediately. Don't make your functions first-class unless you actually want them to be.

Reply:
Ah, I that makes sense - I moved it up.